drgn/helpers/linux/dmesg.py

```diff
@@ -0,0 +1,106 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+"""
+dmesg-vmcore
+-------
+
+The ``drgn.helpers.linux.dmesg`` module provides helpers for reading the
+Linux kernel log.
+"""
+
+from drgn import Program, Object, cast, sizeof
+
+__all__ = ("get_dmesg",)
+
+DESC_SV_BITS = 0
+DESC_FLAGS_SHIFT = 0
+DESC_FLAGS_MASK = 0
+DESC_ID_MASK = 0
+
+def id_incr(counter):
+    return (counter + 1) & DESC_ID_MASK
+
+def record_committed(current_id, state_var):
+    state_desc_id = state_var & DESC_ID_MASK
+    state = 3 & (state_var >> DESC_FLAGS_SHIFT)
+    return (current_id == state_desc_id) and (state == 0x1 or state == 0x2)
+
+def dump_record(prog: Program, desc_ring, desc_ring_count, text_data_ring, text_data_ring_size, ulong_size, current_id):
+    idx = current_id % desc_ring_count
+    text_len = desc_ring.infos[idx].text_len
+    descs = desc_ring.descs[idx]
+
+    if not record_committed(current_id, descs.state_var.counter):
+        return b""
+
+    lpos_begin = descs.text_blk_lpos.begin % text_data_ring_size
+    lpos_next = descs.text_blk_lpos.next % text_data_ring_size
+    lpos_begin += ulong_size
+
+    if lpos_begin == lpos_next:
+        return b""
+    if lpos_begin > lpos_next:
+        lpos_begin = 0
+    if lpos_next - lpos_begin < text_len:
+        text_len = lpos_next - lpos_begin
+    return prog.read(text_data_ring.data + lpos_begin, text_len) + str.encode('\n')
+
+def dump_dmesg_lockless(prog: Program, printk_ringbuffer: Object) -> bytes:
+    global DESC_SV_BITS
+    global DESC_FLAGS_SHIFT
+    global DESC_FLAGS_MASK
+    global DESC_ID_MASK
```

Comment:
Instead of making these global, you can define `dump_record()` and `record_committed()` inside of this function and they'll have these variables available in their scope.